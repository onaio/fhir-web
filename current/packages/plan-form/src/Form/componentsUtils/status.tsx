import React from 'react';
import { Radio, Popconfirm } from 'antd';
import { PlanStatus, planStatusDisplay, status } from '@opensrp/plan-form-core';
import { FormInstance } from 'antd/lib/form';
import { useTranslation } from '../../mls';
import { PlanFormFields } from '../../helpers/types';
import { PopconfirmProps } from 'antd/lib/popconfirm';

interface PlanStatusRendererProps {
  disabledFields: string[];
  disAllowedStatusChoices: string[];
  setFieldsValue: FormInstance['setFieldsValue'];
  assignedJurisdictions?: PlanFormFields['jurisdictions'];
}

/**
 * renders the status fields on the planForm
 * Wraps the radio buttons in a pop up confirmation box, this means
 * we need some custom logic upstream to change how antd form updates
 * values, so that when user does not confirm in the popup, we retain
 * the previous value.
 *
 * @param props - this components props
 */
export const PlanStatusRenderer = (props: PlanStatusRendererProps) => {
  const {
    disabledFields,
    disAllowedStatusChoices,
    setFieldsValue,
    assignedJurisdictions,
    ...rest
  } = props;
  const { t } = useTranslation();

  const popUpConfirmationMessageLookup = {
    [PlanStatus.DRAFT]: t('Are you sure? status will be set to draft'),
    [PlanStatus.ACTIVE]: t("Are you sure? you won't be able to change the status back to draft"),
    [PlanStatus.COMPLETE]: t(
      "Are you sure? you won't be able to change the status for complete plans"
    ),
    [PlanStatus.RETIRED]: t(
      "Are you sure? you won't be able to change the status for retired plans"
    ),
  };

  return (
    <Radio.Group className="plan-form-status" {...rest} disabled={disabledFields.includes(status)}>
      {Object.entries(PlanStatus)
        .filter((e) => !disAllowedStatusChoices.includes(e[1]))
        .map((e) => {
          const titleMessage = () => {
            return <p className="popup-width">{popUpConfirmationMessageLookup[e[1]]}</p>;
          };

          // nominal popConfirmProps
          let popConfirmProps: PopconfirmProps = {
            title: titleMessage,
            onConfirm: () => {
              setFieldsValue({ [status]: e[1] });
            },
            onCancel: () => void 0,
            okText: t('yes'),
            cancelText: t('no'),
          };
          let additionalRadioClassName = '';

          if (e[1] === PlanStatus.ACTIVE && assignedJurisdictions?.length === 0) {
            // change onConfirm to do nothing, and hide cancelButton, change popup title
            popConfirmProps = {
              title: t('Assign jurisdictions to the Plan, to enable activating it'),
              okText: t('OK'),
              cancelButtonProps: {
                style: { display: 'none' },
              },
            };
            // change Radio to be disabled
            additionalRadioClassName = 'disabled';
          }

          return (
            <Popconfirm key={e[0]} {...popConfirmProps}>
              <Radio className={`status-radio ${additionalRadioClassName}`} value={e[1]}>
                {planStatusDisplay[e[1]]}
              </Radio>
            </Popconfirm>
          );
        })}
    </Radio.Group>
  );
};
