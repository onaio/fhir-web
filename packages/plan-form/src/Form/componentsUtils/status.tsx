import React from 'react';
import { Radio, Popconfirm } from 'antd';
import { PlanStatus, planStatusDisplay, status } from '@opensrp/plan-form-core';
import { FormInstance } from 'antd/lib/form';
import {
  NO,
  SETTING_STATUS_TO_ACTIVE,
  SETTING_STATUS_TO_COMPLETE,
  SETTING_STATUS_TO_DRAFT,
  SETTING_STATUS_TO_RETIRED,
  YES,
} from '../../lang';

interface PlanStatusRendererProps {
  disabledFields: string[];
  disAllowedStatusChoices: string[];
  setFieldsValue: FormInstance['setFieldsValue'];
}

const popUpConfirmationMessageLookup = {
  [PlanStatus.DRAFT]: SETTING_STATUS_TO_DRAFT,
  [PlanStatus.ACTIVE]: SETTING_STATUS_TO_ACTIVE,
  [PlanStatus.COMPLETE]: SETTING_STATUS_TO_COMPLETE,
  [PlanStatus.RETIRED]: SETTING_STATUS_TO_RETIRED,
};

/** renders the status fields on the planForm
 * Wraps the radio buttons in a pop up confirmation box, this means
 * we need some custom logic upstream to change how antd form updates
 * values, so that when user does not confirm in the popup, we retain
 * the previous value.
 *
 * @param props - this components props
 */
export const PlanStatusRenderer = (props: PlanStatusRendererProps) => {
  const { disabledFields, disAllowedStatusChoices, setFieldsValue, ...rest } = props;
  return (
    <Radio.Group className="plan-form-status" {...rest} disabled={disabledFields.includes(status)}>
      {Object.entries(PlanStatus)
        .filter((e) => !disAllowedStatusChoices.includes(e[1]))
        .map((e) => {
          const titleMessage = () => {
            return <p className="popup-width">{popUpConfirmationMessageLookup[e[1]]}</p>;
          };

          return (
            <Popconfirm
              key={e[0]}
              title={titleMessage}
              onConfirm={() => {
                setFieldsValue({ [status]: e[1] });
              }}
              onCancel={() => void 0}
              okText={YES}
              cancelText={NO}
            >
              <Radio className="status-radio" value={e[1]}>
                {planStatusDisplay[e[1]]}
              </Radio>
            </Popconfirm>
          );
        })}
    </Radio.Group>
  );
};
