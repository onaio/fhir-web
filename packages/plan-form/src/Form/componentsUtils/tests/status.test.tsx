import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { PlanStatusRenderer } from '../status';

describe('status Renderer', () => {
  it('renderers correctly', () => {
    const props = { disabledFields: [], disAllowedStatusChoices: [], setFieldsValue: jest.fn() };
    const wrapper = mount(<PlanStatusRenderer {...props} />);

    expect(wrapper.find('.status-radio')).toHaveLength(8);
    wrapper.find('Radio.status-radio').forEach((radio) => {
      expect(radio.find('label').text()).toMatchSnapshot('radio button label');
      expect(toJson(radio.find('input'))).toMatchSnapshot('radio button input');
    });
  });

  it('shows a popup when changing to complete', () => {
    const container = document.createElement('div');
    document.body.appendChild(container);
    const setFieldMock = jest.fn();
    const props = { disabledFields: [], disAllowedStatusChoices: [], setFieldsValue: setFieldMock };
    const wrapper = mount(<PlanStatusRenderer {...props} />, { attachTo: container });

    // attempt simulating confirm on complete radio button
    wrapper.find('input[value="complete"]').simulate('click');
    // look for popup to confirm
    expect(wrapper.find('.ant-popover-content').text()).toMatchInlineSnapshot(
      `"Are you sure, you won't be able to change the status for complete plansnoyes"`
    );

    // simulate click on deny button
    expect(wrapper.find('button').first().text()).toMatchInlineSnapshot(`"no"`);
    wrapper.find('button').first().simulate('click');
    // nothing should change
    expect(setFieldMock).not.toHaveBeenCalled();

    // simulate click on confirm button
    expect(wrapper.find('button').last().text()).toMatchInlineSnapshot(`"yes"`);
    wrapper.find('button').last().simulate('click');
    // nothing should change
    expect(setFieldMock).toHaveBeenCalledWith({ status: 'complete' });
    wrapper.unmount();
  });

  it('shows a popup when clicking on active', () => {
    const container = document.createElement('div');
    document.body.appendChild(container);
    const setFieldMock = jest.fn();
    const props = { disabledFields: [], disAllowedStatusChoices: [], setFieldsValue: setFieldMock };
    const wrapper = mount(<PlanStatusRenderer {...props} />, { attachTo: container });

    // attempt simulating confirm on active radio button
    wrapper.find('input[value="active"]').simulate('click');
    // look for popup to confirm
    expect(wrapper.find('.ant-popover-content').text()).toMatchInlineSnapshot(
      `"Are you sure, you won't be able to change the status back to draftnoyes"`
    );

    // simulate click on deny button
    wrapper.find('button').first().simulate('click');
    // nothing should change
    expect(setFieldMock).not.toHaveBeenCalled();

    // simulate click on confirm button
    wrapper.find('button').last().simulate('click');
    // nothing should change
    expect(setFieldMock).toHaveBeenCalledWith({ status: 'active' });
  });

  it('shows a popup when clicking on draft', () => {
    const container = document.createElement('div');
    document.body.appendChild(container);
    const setFieldMock = jest.fn();
    const props = { disabledFields: [], disAllowedStatusChoices: [], setFieldsValue: setFieldMock };
    const wrapper = mount(<PlanStatusRenderer {...props} />, { attachTo: container });

    // attempt simulating confirm on active radio button
    wrapper.find('input[value="draft"]').simulate('click');
    // look for popup to confirm
    expect(wrapper.find('.ant-popover-content').text()).toMatchInlineSnapshot(
      `"Are you sure, status will be set to draftnoyes"`
    );

    // simulate click on deny button
    wrapper.find('button').first().simulate('click');
    // nothing should change
    expect(setFieldMock).not.toHaveBeenCalled();

    // simulate click on confirm button
    wrapper.find('button').last().simulate('click');
    // nothing should change
    expect(setFieldMock).toHaveBeenCalledWith({ status: 'draft' });
  });

  it('shows a popup when clicking on retired', () => {
    const container = document.createElement('div');
    document.body.appendChild(container);
    const setFieldMock = jest.fn();
    const props = { disabledFields: [], disAllowedStatusChoices: [], setFieldsValue: setFieldMock };
    const wrapper = mount(<PlanStatusRenderer {...props} />, { attachTo: container });

    // attempt simulating confirm on active radio button
    wrapper.find('input[value="retired"]').simulate('click');
    // look for popup to confirm
    expect(wrapper.find('.ant-popover-content').text()).toMatchInlineSnapshot(
      `"Are you sure, you won't be able to change the status for retired plansnoyes"`
    );

    // simulate click on deny button
    wrapper.find('button').first().simulate('click');
    // nothing should change
    expect(setFieldMock).not.toHaveBeenCalled();

    // simulate click on confirm button
    wrapper.find('button').last().simulate('click');
    // nothing should change
    expect(setFieldMock).toHaveBeenCalledWith({ status: 'retired' });
  });
});
