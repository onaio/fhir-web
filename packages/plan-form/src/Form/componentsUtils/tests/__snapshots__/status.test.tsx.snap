// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`status Renderer renderers correctly: radio button input 1`] = `
<input
  checked={false}
  className="ant-radio-input"
  disabled={false}
  onBlur={[Function]}
  onChange={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyPress={[Function]}
  onKeyUp={[Function]}
  type="radio"
  value="active"
/>
`;

exports[`status Renderer renderers correctly: radio button input 2`] = `
<input
  checked={false}
  className="ant-radio-input"
  disabled={false}
  onBlur={[Function]}
  onChange={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyPress={[Function]}
  onKeyUp={[Function]}
  type="radio"
  value="complete"
/>
`;

exports[`status Renderer renderers correctly: radio button input 3`] = `
<input
  checked={false}
  className="ant-radio-input"
  disabled={false}
  onBlur={[Function]}
  onChange={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyPress={[Function]}
  onKeyUp={[Function]}
  type="radio"
  value="draft"
/>
`;

exports[`status Renderer renderers correctly: radio button input 4`] = `
<input
  checked={false}
  className="ant-radio-input"
  disabled={false}
  onBlur={[Function]}
  onChange={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyPress={[Function]}
  onKeyUp={[Function]}
  type="radio"
  value="retired"
/>
`;

exports[`status Renderer renderers correctly: radio button label 1`] = `"active"`;

exports[`status Renderer renderers correctly: radio button label 2`] = `"complete"`;

exports[`status Renderer renderers correctly: radio button label 3`] = `"draft"`;

exports[`status Renderer renderers correctly: radio button label 4`] = `"retired"`;
